### PR TITLE
plugin for NFS exporting local filesystems

### PR DESCRIPTION
**Pull request by jumitche**
First cut of a plugin that allows for the exporting of local filesystems by nfs, this can be used to add the missing export functionality to hpsa/megaraid plugins, as well as being used on its own.

Note: Trusted repo added, tests will run.

### CODE REVIEW

File: doc/man/nfs_lsmplugin.1.in [hunk 1]
```diff
@@ -0,0 +1,30 @@
+.TH nfs_lsmplugin "1" "February 2017" "nfs_lsmplugin @VERSION@" "libStorageMgmt"
+.SH NAME
+nfs_lsmplugin -- LibstorageMgmt local filesystem exports plugin
```
Comment (cathay4t) [inline doc/man/nfs_lsmplugin.1.in]: How about `local network file system plugin`?

File: doc/man/nfs_lsmplugin.1.in [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    nfs://
+
+.fi
+No password is required for this plugin.
+No URI parameters are supported by this plugin.
+
+.SH FIREWALL RULES
+This plugin requires no network access.
+
```
Comment (cathay4t) [inline doc/man/nfs_lsmplugin.1.in]: Please also include a section:

```man
.SH ROOT PRIVILEGE
This plugin requires both \fBlsmd\fR daemon and API client running as root
user. Please check manpage \fIlsmd.conf (5)\fR for detail.
```

File: packaging/libstoragemgmt.spec.in [hunk 3]
```diff
@@ -805,6 +809,10 @@ fi
 %{udev_dir}/udev/rules.d/90-scsi-ua.rules
 
 %changelog
+* Mon Feb 6 2017 Justin Mitchell <jumitche@redhat.com> 1.3.5-1
```
Comment (cathay4t) [inline packaging/libstoragemgmt.spec.in]: Current master branch is marked as version 1.4.0.

**Comment by cathay4t** [comment]
Haven't read through the C code yet, but I slightly dislike the idea of invoke CLI command in C file.

**Comment by jumitche** [comment]
> Haven't read through the C code yet, but I slightly dislike the idea of invoke CLI command in C file.
It is not ideal, I am still looking for a better way to do it that doesn't involve reimplementing all of exportfs, any suggestions for a library or other simple way gratefully received.

**Comment by tasleson** [comment]
Well I'm kind of stoked that someone has used the C API to create a plugin, but yeah writing C to call a command line tool seems counter productive.  Any particular reason when the plugin wasn't written in python?  Seems like that would have been a better fit for exec'ing command line tools.

**Comment by jumitche** [comment]
In this case it doesn't need or attempt to parse anything from the commandline, it just has to run the program, which from asking around is the only sane way to update the exports.
I had started out with the idea that this would be shared code and that perhaps C would be more convenient for other plugins to call, but in hindsight maybe it would have been better in python

**Comment by cathay4t** [comment]
Concerns:
 * `strerror()` is not thread safe, please use `strerror_r()`.
 * `list_exports()` only show exports defined in file `/etc/exports.d/libstoragemgmt.exports`, it might not contain all entries the kernel exposed. Do we have any better way like checking kernel kernel export table(something like `/proc/fs/nfsd`) or `exportfs -s`?
 * For `fs_list()`, `getmntent()` is better than do it by yourself.
 * Parsing files or cli output is much easier in python than C.

**Comment by cathay4t** [comment]
For sharing code between projects, I would suggest to create a standalone library like `libnfsd` and `python-libnfsd` which libstoragemgmt nfs plugin could disburden.

**Comment by jumitche** [comment]
> strerror() is not thread safe, please use strerror_r(

Oh, i thought these plugins were single threaded stand alone processes, i will change it.

> list_exports() only show exports defined in file /etc/exports.d/libstoragemgmt.exports, it might not contain all entries the kernel exposed. Do we have any better way like checking kernel kernel export table(something like /proc/fs/nfsd) or exportfs -s?

This was a choice, i originally started out by parsing /var/lib/nfs/etab which is what exportfs writes to, and lists all of the current exports. but, it would then mean listing exports that we couldnt delete/modify easily, because they could be coming from hand crafted entries anywhere in /etc/exports or /etc/exports.d/* and those files allow for some quite complex constructions. So rather than wade in and start messing up other peoples configs i thought it better to restrict it to only exports it has created and can manage itself.

> For fs_list(), getmntent() is better than do it by yourself.

I missed that one, thanks.

**Comment by cathay4t** [comment]
@jumitche If we are only support NFS exports in `/etc/exports.d/libstoragemgmt.exports`, please mention that in manpage of the nfs plugin to stop people asking around.

**Comment by cathay4t** [comment]
Yes, plugin is in their own process for each lsm connection. But we don't mention that in any document and open to changes if needed in the future. So it's better to make sure plugin is thread safe.

**Comment by tasleson** [comment]
@jumitche 
```
make[3]: Entering directory '/tmp/lsm_ci/src/13212/libstoragemgmt/plugin/nfs'
  CC       nfs_lsmplugin.o
nfs_lsmplugin.c: In function ‘path_to_fsid’:
nfs_lsmplugin.c:205:27: error: format ‘%llx’ expects argument of type ‘long long unsigned int’, but argument 3 has type ‘long unsigned int’ [-Werror=format=]
     if (asprintf(&answer, "%" PRIx64, st.f_fsid)<0) {
                           ^
nfs_lsmplugin.c: In function ‘fs_list’:
nfs_lsmplugin.c:987:29: error: format ‘%llx’ expects argument of type ‘long long unsigned int’, but argument 3 has type ‘long unsigned int’ [-Werror=format=]
         if (asprintf(&fsid, "%" PRIx64, st.f_fsid) < 0) {
```

**Comment by jumitche** [comment]
@tasleson  thanks, that one didnt even show up with a -pedantic on my system

**Comment by jumitche** [comment]
I am trying to rework this to be mostly python, there are a few bits that will still be done in C as there seems to be no sane way to do them in python like getting the fsid value, they are generic to local filesystems so i will put them with the existing local disk code

**Pull request closed without merging**